### ISSUE TITLE
tf.keras.losses.SparseCategoricalCrossentropy within the multi-GPU strategy.scope() generates NAN value

### ISSUE THREAD

**Issue by YINWIZHANG** [tensorflow/tensorflow#36845]
My input is `rgb` and `depth` are all `[480,640,3]` images, and the `labels` are `[480,640,1]`. When I use` losses.SparseCategoricalCrossentropy()` in single GPU, it works fine and gives a single loss value. However, when I use it under multi-gpu case, it gives wrong values. Any one can solve this issue?
```
if single_GPU == 'True':
    print('Train on single GPUs...')
    train_ds = train.batch(global_batch_size)

    with tf.device('device:GPU:0'):
        network = resnet_fcn1()
        optimizer = optimizers.SGD(0.001)
        L = losses.SparseCategoricalCrossentropy(from_logits=True)

    for epoch in range(total_epoch):
        batch = 0
        for rgb, depth, label in train_ds:
            with tf.GradientTape() as tape:
                pred = network([rgb, depth])
                loss = L(label, pred)
            print(template.format(epoch, batch, loss))
            batch += 1
            grads = tape.gradient(loss, network.trainable_variables)
            optimizer.apply_gradients((zip(grads, network.trainable_variables)))
            if batch == 1:
                break
else:
    print('Train on multiple GPUs...')

    BATCH_SIZE_PER_REPLICA = 2
    strategy = tf.distribute.MirroredStrategy()

    GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync

    train_ds = train.batch(GLOBAL_BATCH_SIZE)
    train_ds = strategy.experimental_distribute_dataset(train_ds)
    
    with strategy.scope():
        loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True, reduction = tf.keras.losses.Reduction.NONE)
        def compute_loss(labels, predictions):
           per_example_loss = loss_object(labels, predictions)
           return tf.nn.compute_average_loss(per_example_loss, global_batch_size=GLOBAL_BATCH_SIZE)


    with strategy.scope():
        network = resnet_fcn1()
        optimizer = optimizers.SGD()

    with strategy.scope():
        def train_step(inputs):
            rgb, depth, labels = inputs
            with tf.GradientTape() as tape:
                predictions = network([rgb, depth])
                loss = compute_loss(labels, predictions)
            gradients = tape.gradient(loss, network.trainable_variables)
            optimizer.apply_gradients(zip(gradients, network.trainable_variables))
            return loss

    with strategy.scope():
        @tf.function
        def distributed_train_step(dataset_inputs):
            per_replica_losses = strategy.experimental_run_v2(train_step, args=(dataset_inputs, ))
            return strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)

        for epoch in range(total_epoch):
            batch = 0
            for inputs in train_ds:
                loss = distributed_train_step(inputs)
                print(template.format(epoch, batch, loss))
                batch += 1
```

**Comment by YINWIZHANG** [comment]
In the single-gpu case, I have
```
epoch:0, batch:0, loss:1.0988747737
epoch:0, batch:1, loss:0.9993420934
epoch:0, batch:2, loss:0.9823420342
```

In the multi-gpu case, I have the result
```
epoch:0, batch:0, loss:11783.943359375
epoch:0, batch:1, loss:inf
epoch:0, batch:2, loss:inf
epoch:0, batch:3, loss:inf
```

**Comment by ravikyram** [comment]
@YINWIZHANG 

Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version. 
Request you to fill the [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).Thanks!

**Comment by YINWIZHANG** [comment]
> @YINWIZHANG
> 
> Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version.
> Request you to fill the [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).Thanks!

I am using ubuntu 18.04 and tensorflow 2.0.0. My PC is a Lambda stack which has 4 RTX8000 GPUs.

**Comment by caisq** [comment]
@YINWIZHANG In order to help us investigate this issue, can you provide the following extra bits of information:
- Details of your `resnet_fcn1`
- Details of your training dataset, `train_ds`.

If there can be actual code that forms a minimal reproduction case it'll be great.

**Comment by YINWIZHANG** [comment]
> @YINWIZHANG In order to help us investigate this issue, can you provide the following extra bits of information:
> 
>     * Details of your `resnet_fcn1`
> 
>     * Details of your training dataset, `train_ds`.
> 
> 
> If there can be actual code that forms a minimal reproduction case it'll be great.
Hi, here is the minimal reproduction case.

`import tensorflow as tf
from tensorflow.keras import optimizers, losses, layers, applications, Input, Model
import os
import numpy as np
import matplotlib.pyplot as plt
import time


total_epoch = 1
BATCH_SIZE_PER_REPLICA = 1
strategy = tf.distribute.MirroredStrategy()
template = 'epoch: {}, batch: {}, loss: {}'
GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync


def resnet_fcn2(nclass = 3):
    print('import resetnet101...')
    #define a resnet 101 for rgb image
    rgb_model = applications.ResNet101(weights = 'imagenet')
    layer_list = [layer.output for layer in rgb_model.layers]
    rgb_model = tf.keras.Model(inputs = rgb_model.input, outputs = layer_list[312])
    #define a resnet 101 for the depth image
    depth_model = applications.ResNet101(weights='imagenet')
    layer_list = [layer.output for layer in depth_model.layers]
    depth_model = tf.keras.Model(inputs=depth_model.input, outputs=layer_list[312])
    #rename them to avoid error
    rgb_model._name = 'rgb_Model'
    depth_model._name = 'Depth_Model'
    #define input
    rgb_in = Input(shape=[480, 640, 3], name = 'rgb')
    depth_in = Input(shape=[480, 640, 3], name='depth')
    #get output
    rgb_out = rgb_model(rgb_in)
    depth_out = depth_model(depth_in)
    #concatenate two outputs
    output = layers.concatenate([rgb_out, depth_out], axis=3)
    #pass the output through several convolutional layers
    output = layers.Conv2D(filters = 512, kernel_size = (1, 1), strides = (1, 1))(output)
    output = layers.Conv2D(filters = 128, kernel_size = (1, 1), strides = (1, 1))(output)
    output = layers.Conv2D(filters = nclass, kernel_size = (1, 1), strides = (1, 1))(output)
    #upsample the output
    output = layers.UpSampling2D(size=(2, 2), interpolation='bilinear')(output)  #the output size should be 80*60
    model = Model([rgb_in, depth_in], output)
    print('network building, complete...')
    return model


rgb = tf.ones([100, 480, 640, 3])   #input 1
depth = tf.ones([100, 480, 640, 3])  #input 2
label = tf.ones([100, 60, 80])  #label

train = tf.data.Dataset.from_tensor_slices((rgb, depth, label))
train_ds = train.batch(GLOBAL_BATCH_SIZE)
####################################################training using single GPU####################################
print('Train on single GPUs...')
with tf.device('device:GPU:0'):
    network = resnet_fcn2() #load the model
    optimizer = optimizers.SGD(0.001)
    L = losses.SparseCategoricalCrossentropy(from_logits=True) #define the loss

for epoch in range(total_epoch):
    batch = 0
    for rgb, depth, label in train_ds:
        with tf.GradientTape() as tape:
            pred = network([rgb, depth])
            loss = L(label, pred)
        print(template.format(epoch, batch, loss))
        batch += 1
        grads = tape.gradient(loss, network.trainable_variables)
        optimizer.apply_gradients((zip(grads, network.trainable_variables)))
#############################################################################################################
print('Train on multiple GPUs...')
train_ds = strategy.experimental_distribute_dataset(train_ds)

loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True, reduction = tf.keras.losses.Reduction.NONE)
def compute_loss(labels, predictions):
    per_example_loss = loss_object(labels, predictions)
    return tf.nn.compute_average_loss(per_example_loss, global_batch_size=GLOBAL_BATCH_SIZE)

with strategy.scope():
    network = resnet_fcn2()
    optimizer = optimizers.SGD()

with strategy.scope():
    def train_step(inputs):
        rgb, depth, labels = inputs
        with tf.GradientTape() as tape:
            predictions = network([rgb, depth])
            loss = compute_loss(labels, predictions)
        gradients = tape.gradient(loss, network.trainable_variables)
        optimizer.apply_gradients(zip(gradients, network.trainable_variables))
        return loss

with strategy.scope():
    @tf.function
    def distributed_train_step(dataset_inputs):
        per_replica_losses = strategy.experimental_run_v2(train_step, args=(dataset_inputs, ))
        return strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)

    template = 'epoch: {}, batch: {}, loss: {}'
    for epoch in range(total_epoch):
        batch = 0
        for inputs in train_ds:
            loss = distributed_train_step(inputs)
            print(template.format(epoch, batch, loss))
            batch += 1`

**Comment by YINWIZHANG** [comment]
> @YINWIZHANG In order to help us investigate this issue, can you provide the following extra bits of information:
> 
>     * Details of your `resnet_fcn1`
> 
>     * Details of your training dataset, `train_ds`.
> 
> 
> If there can be actual code that forms a minimal reproduction case it'll be great.
Here is the reproduction case.
`import tensorflow as tf
from tensorflow.keras import optimizers, losses, layers, applications, Input, Model
import os
import numpy as np
import matplotlib.pyplot as plt
import time


total_epoch = 1
BATCH_SIZE_PER_REPLICA = 1
strategy = tf.distribute.MirroredStrategy()
template = 'epoch: {}, batch: {}, loss: {}'
GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync


def resnet_fcn2(nclass = 3):
    print('import resetnet101...')
    #define a resnet 101 for rgb image
    rgb_model = applications.ResNet101(weights = 'imagenet')
    layer_list = [layer.output for layer in rgb_model.layers]
    rgb_model = tf.keras.Model(inputs = rgb_model.input, outputs = layer_list[312])
    #define a resnet 101 for the depth image
    depth_model = applications.ResNet101(weights='imagenet')
    layer_list = [layer.output for layer in depth_model.layers]
    depth_model = tf.keras.Model(inputs=depth_model.input, outputs=layer_list[312])
    #rename them to avoid error
    rgb_model._name = 'rgb_Model'
    depth_model._name = 'Depth_Model'
    #define input
    rgb_in = Input(shape=[480, 640, 3], name = 'rgb')
    depth_in = Input(shape=[480, 640, 3], name='depth')
    #get output
    rgb_out = rgb_model(rgb_in)
    depth_out = depth_model(depth_in)
    #concatenate two outputs
    output = layers.concatenate([rgb_out, depth_out], axis=3)
    #pass the output through several convolutional layers
    output = layers.Conv2D(filters = 512, kernel_size = (1, 1), strides = (1, 1))(output)
    output = layers.Conv2D(filters = 128, kernel_size = (1, 1), strides = (1, 1))(output)
    output = layers.Conv2D(filters = nclass, kernel_size = (1, 1), strides = (1, 1))(output)
    #upsample the output
    output = layers.UpSampling2D(size=(2, 2), interpolation='bilinear')(output)  #the output size should be 80*60
    model = Model([rgb_in, depth_in], output)
    print('network building, complete...')
    return model


rgb = tf.ones([100, 480, 640, 3])   #input 1
depth = tf.ones([100, 480, 640, 3])  #input 2
label = tf.ones([100, 60, 80])  #label

train = tf.data.Dataset.from_tensor_slices((rgb, depth, label))
train_ds = train.batch(GLOBAL_BATCH_SIZE)
####################################################training using single GPU####################################
print('Train on single GPUs...')
with tf.device('device:GPU:0'):
    network = resnet_fcn2() #load the model
    optimizer = optimizers.SGD(0.001)
    L = losses.SparseCategoricalCrossentropy(from_logits=True) #define the loss

for epoch in range(total_epoch):
    batch = 0
    for rgb, depth, label in train_ds:
        with tf.GradientTape() as tape:
            pred = network([rgb, depth])
            loss = L(label, pred)
        print(template.format(epoch, batch, loss))
        batch += 1
        grads = tape.gradient(loss, network.trainable_variables)
        optimizer.apply_gradients((zip(grads, network.trainable_variables)))
#############################################################################################################
print('Train on multiple GPUs...')
train_ds = strategy.experimental_distribute_dataset(train_ds)

loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True, reduction = tf.keras.losses.Reduction.NONE)
def compute_loss(labels, predictions):
    per_example_loss = loss_object(labels, predictions)
    return tf.nn.compute_average_loss(per_example_loss, global_batch_size=GLOBAL_BATCH_SIZE)

with strategy.scope():
    network = resnet_fcn2()
    optimizer = optimizers.SGD()

with strategy.scope():
    def train_step(inputs):
        rgb, depth, labels = inputs
        with tf.GradientTape() as tape:
            predictions = network([rgb, depth])
            loss = compute_loss(labels, predictions)
        gradients = tape.gradient(loss, network.trainable_variables)
        optimizer.apply_gradients(zip(gradients, network.trainable_variables))
        return loss

with strategy.scope():
    @tf.function
    def distributed_train_step(dataset_inputs):
        per_replica_losses = strategy.experimental_run_v2(train_step, args=(dataset_inputs, ))
        return strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)

    template = 'epoch: {}, batch: {}, loss: {}'
    for epoch in range(total_epoch):
        batch = 0
        for inputs in train_ds:
            loss = distributed_train_step(inputs)
            print(template.format(epoch, batch, loss))
            batch += 1`

**Comment by m4tts** [comment]
@YINWIZHANG I experienced a similar problem. At first the loss exploded, after some day the loss stalled and at some point, everything just started crashing. In the end is was a defective GPU.
It doesn't mean you have the same problem, but maybe those tipps will help you narrow it down :

- Does every GPU work fine? Did you try using `tf.device('device:GPU:1'):` or 2,3. ?

- Test with [gpu-burn](https://github.com/wilicc/gpu-burn) to check, if all GPUs work properly.

**Comment by YINWIZHANG** [comment]
> @YINWIZHANG I experienced a similar problem. At first the loss exploded, after some day the loss stalled and at some point, everything just started crashing. In the end is was a defective GPU.
> It doesn't mean you have the same problem, but maybe those tipps will help you narrow it down :
> 
>     * Does every GPU work fine? Did you try using `tf.device('device:GPU:1'):` or 2,3. ?
> 
>     * Test with [gpu-burn](https://github.com/wilicc/gpu-burn) to check, if all GPUs work properly.

Thanks for your comments! It turns out that I have all my GPUs working fine. A really strange thing is that, even when I use above multiGPU code in a machine with single GPU, the loss value explodes.

**Comment by caisq** [comment]
@YINWIZHANG 

Can you try modifying your loss function under the MirroredStrategy as follows:

```py
def compute_loss(labels, predictions):
  per_example_loss = loss_object(labels, predictions)
  per_example_loss /= tf.cast(tf.size(labels), tf.float32)  # <--- See if this fixes the NaNs.
  return tf.nn.compute_average_loss(
      per_example_loss, global_batch_size=GLOBAL_BATCH_SIZE)
```

**Comment by YINWIZHANG** [comment]
> @YINWIZHANG
> 
> Can you try modifying your loss function under the MirroredStrategy as follows:
> 
> ```python
> def compute_loss(labels, predictions):
>   per_example_loss = loss_object(labels, predictions)
>   per_example_loss /= tf.cast(
>       tf.shape(labels)[1] * tf.shape(labels)[2], tf.float32)  # <--- See if this fixes the NaNs.
>   return tf.nn.compute_average_loss(
>       per_example_loss, global_batch_size=GLOBAL_BATCH_SIZE)
> ```

Thanks! It solves the problem!
